### ISSUE TITLE
Better implementation of metrics in NN architectures

### ISSUE THREAD

**Issue by nsorros** [wellcometrust/WellcomeML#42]
Instead of a callback pass a custom function that works on calculating the metrics of interest. This does not work if we naively pass f1_score or even a partial f1_score with the average instantiated for different reasons.

**Comment by nsorros** [comment]
@aCampello actually there is also a ticket for the metrics, want to write down what you plan to implement if you want to work on this?

**Comment by aCampello** [comment]
Yes. I was planning to:

- Define metrics as their own custom metric as in https://keras.io/metrics/
- Add a callback to add metrics to the history of model at the end of each mini-batch (instead of epoch), so we could potentially plot a history by calling `classifier.model.history`.

Something that will probably look like

`def get_keras_custom_metrics(metric_names)`

and the possible metric names are `f1_score`, `precision`, `recall`, etc, and this will return a Keras-compatible metric. So from the classifiers you could pass the parameter
`metrics=get_keras_custom_metrics(['f1_score', 'precision', 'recall'])`

You could even add built-in metrics with this to the model like:
`metrics += ['sparse_categorical_cross_entropy']`

Is that similar to what you thought?

**Comment by nsorros** [comment]
I thought each metric would have its own function and get called from metrics something like `metrics=[f1_score, prediction_score, recall_score]`

**Comment by nsorros** [comment]
if they are working out of the box 100%
